drivers/flash/spi_flash_at45.c

```diff
@@ -0,0 +1,685 @@
+
+
+#define DEF_BUF_SET(_name, _buf_array) \
+	const struct spi_buf_set _name = { \
+		.buffers = _buf_array, \
+		.count   = ARRAY_SIZE(_buf_array), \
+	}
+
+struct spi_flash_at45_data {
+	struct device *spi;
+	struct spi_cs_control spi_cs;
+	struct k_sem lock;
+#if IS_ENABLED(CONFIG_DEVICE_POWER_MANAGEMENT)
+	u32_t pm_state;
+#endif
+};
+
+struct spi_flash_at45_config {
+	const char *spi_bus;
+	struct spi_config spi_cfg;
+	const char *cs_gpio;
+	int cs_pin;
+#if IS_ENABLED(CONFIG_FLASH_PAGE_LAYOUT)
+	struct flash_pages_layout pages_layout;
+#endif
+	u32_t chip_size;
+	u16_t block_size;
+	u16_t page_size;
+	u16_t t_enter_dpd; /* in microseconds */
+	u16_t t_exit_dpd;  /* in microseconds */
+	bool use_udpd;
+	u8_t jedec_id[3];
+};
+
+static struct spi_flash_at45_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static const struct spi_flash_at45_config *get_dev_config(struct device *dev)
+{
+	return dev->config->config_info;
+}
+
+static void acquire(struct device *dev)
+{
+	k_sem_take(&get_dev_data(dev)->lock, K_FOREVER);
+}
+
+static void release(struct device *dev)
+{
+	k_sem_give(&get_dev_data(dev)->lock);
+}
+
+static int check_jedec_id(struct device *dev)
+{
+	const struct spi_flash_at45_config *cfg = get_dev_config(dev);
+	int err;
+	u8_t const *expected_id = cfg->jedec_id;
+	u8_t read_id[sizeof(cfg->jedec_id)];
+	const u8_t opcode = CMD_READ_ID;
+	const struct spi_buf tx_buf[] = {
+		{
+			.buf = (void *)&opcode,
+			.len = sizeof(opcode),
+		}
+	};
+	const struct spi_buf rx_buf[] = {
+		{
+			.len = sizeof(opcode),
+		},
+		{
+			.buf = read_id,
+			.len = sizeof(read_id),
+		}
+	};
+	DEF_BUF_SET(tx_buf_set, tx_buf);
+	DEF_BUF_SET(rx_buf_set, rx_buf);
+
+	err = spi_transceive(get_dev_data(dev)->spi,
+			     &cfg->spi_cfg,
+			     &tx_buf_set, &rx_buf_set);
+	if (err != 0) {
+		LOG_ERR("SPI transaction failed with code: %d/%u",
+			err, __LINE__);
+		return -EIO;
+	}
+
+	if (memcmp(expected_id, read_id, sizeof(read_id)) != 0) {
+		LOG_ERR("Wrong JEDEC ID: %02X %02X %02X, "
+			"expected: %02X %02X %02X",
+			read_id[0], read_id[1], read_id[2],
+			expected_id[0], expected_id[1], expected_id[2]);
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+/*
+ * Reads 2-byte Status Register:
+ * - Byte 0 to LSB
+ * - Byte 1 to MSB
+ * of the pointed parameter.
+ */
+static int read_status_register(struct device *dev, u16_t *status)
+{
+	int err;
+	const u8_t opcode = CMD_READ_STATUS;
+	const struct spi_buf tx_buf[] = {
+		{
+			.buf = (void *)&opcode,
+			.len = sizeof(opcode),
+		}
+	};
+	const struct spi_buf rx_buf[] = {
+		{
+			.len = sizeof(opcode),
+		},
+		{
+			.buf = status,
+			.len = sizeof(u16_t),
+		}
+	};
+	DEF_BUF_SET(tx_buf_set, tx_buf);
+	DEF_BUF_SET(rx_buf_set, rx_buf);
+
+	err = spi_transceive(get_dev_data(dev)->spi,
+			     &get_dev_config(dev)->spi_cfg,
+			     &tx_buf_set, &rx_buf_set);
+	if (err != 0) {
+		LOG_ERR("SPI transaction failed with code: %d/%u",
+			err, __LINE__);
+	}
+
```

Comment:
I think there should be a `*status = sys_le16_to_cpu(*status)` here on success to put the status register into host byte order.

Reply:
Good point. Added.